### ISSUE TITLE
wazuh-remoted: fix remoted_module_utest ASan failures

### ISSUE THREAD

**Issue by matigarciadev** [wazuh/wazuh#38741]
## Summary

`remoted_module_utest` has failed in CI under AddressSanitizer with two distinct, unrelated memory-safety bugs (different ASan checks, different root causes, different files).

---

## Finding 1 — `FakeUdsServer` detaches per-connection threads → **stack-use-after-return**

CI failure, `WazuhDBClientTest.SuccessfulRoundTripObservesLatencyOnce` (2026-08-27T22:03:01Z run):

```
==6965==ERROR: AddressSanitizer: stack-use-after-return on address 0x7f8296a93590
READ of size 1 at 0x7f8296a93590 thread T2268
    #0 std::atomic<bool>::load(std::memory_order) const
    #2 remoted::test::FakeUdsServer::serveConnection(int) .../fakeUdsServer.hpp:157
    #3 remoted::test::FakeUdsServer::acceptLoop()::{lambda()#1}::operator()() const .../fakeUdsServer.hpp:151
...
Address 0x7f8296a93590 is located in stack of thread T0 at offset 1424 in frame
    #0 WazuhDBClientTest_QueryTimeoutFailsWithTimeoutAndIncrementsMetric_Test::TestBody() .../wazuhDBClient_test.cpp:441
    [1344, 1440) 'server' (line 443) <== Memory access at offset 1424 is inside this variable
Thread T2268 created by T2263 (the FakeUdsServer::acceptLoop thread) here:
    std::thread::thread<...acceptLoop()::{lambda()#1}...>
Thread T2263 created by T0 here:
    std::thread::thread<...FakeUdsServer::FakeUdsServer(...)::{lambda()#1}...>
```

Note the failing read is `m_stopping.load()` at `fakeUdsServer.hpp:157`, called from a `serveConnection` thread that `acceptLoop()` spawned — and that thread is reading into the stack frame of a *different, already-returned* test (`QueryTimeoutFailsWithTimeoutAndIncrementsMetric`'s `server` local), while the currently-running test is `SuccessfulRoundTripObservesLatencyOnce`. That cross-test detail is itself proof of the bug: the detached thread outlived its own test entirely and was still alive when the *next* test's stack reused that memory.

### Root cause

- `FakeUdsServer::acceptLoop()` spawns `std::thread([this, fd] { serveConnection(fd); }).detach()` for every accepted connection.
- `FakeUdsServer::stop()`/`~FakeUdsServer()` only join `m_thread` (the accept loop); they have no reference to detached connection threads and cannot wait for them.
- `FakeUdsServer` is always a stack-local in the test body. A connection thread still blocked in `read()`/`readAll()` at teardown keeps dereferencing `this` after the test function has returned and the stack has been reused by a later test.

---

## Finding 2 — `RestinioHttpServer` shutdown races RESTinio's connection-count limiter → **heap-use-after-free**

CI failure, `AdminServerTest.AdminBindFailureOnlyWarnsAndTheModuleKeepsServing` (2026-08-27T16:39:08Z run):

```
==6814==ERROR: AddressSanitizer: heap-use-after-free on address 0x516000013768
READ of size 8 at 0x516000013768 thread T0
    #0 restinio::connection_count_limits::impl::actual_limiter_t<std::mutex>::decrement_parallel_connections()::{lambda()#1}::operator()() const  connection_count_limiter.hpp:212
    #2 connection_lifetime_monitor_t<...>::~connection_lifetime_monitor_t()  connection_count_limiter.hpp:389
    #3 ~connection_t  connection.hpp:419
    ...
    #33 remoted::http::RestinioHttpServer::stop()  RestinioHttpServer.cpp:1413
    #44 RemotedModuleFacade::stop()  remotedModuleFacade.hpp:257
    #46 remoted_module_stop  remotedModule.cpp:73
    #47 AdminServerTest_AdminBindFailureOnlyWarnsAndTheModuleKeepsServing_Test::TestBody()  adminServer_test.cpp:308

0x516000013768 is located 488 bytes inside of 528-byte region [0x516000013580,0x516000013790)
freed by thread T0 here:
    ... (shared_ptr<_Sp_counted_ptr_inplace<acceptor_t<ServerTraits>>>::_M_dispose(), i.e. the acceptor's last owner going away during io_context shutdown)
previously allocated by thread T0 here:
    ... make_shared<acceptor_t<ServerTraits>>  http_server_t::http_server_t()  http_server.hpp:218
```

### Root cause

- `RestinioHttpServer` enables RESTinio's connection-count limiter (`use_connection_count_limiter = true`, `RestinioHttpServer.cpp:345`) and `separate_accept_and_create_connect(true)` (`RestinioHttpServer.cpp:1339`).
- The limiter is embedded inside `acceptor_t`; every connection only holds a **raw pointer** back to it via `connection_lifetime_monitor_t` (`external/restinio/dev/restinio/connection_count_limiter.hpp`) — by RESTinio's own contract, the acceptor must outlive every connection referencing it.
- On accept, RESTinio increments the limiter immediately and, because `separate_accept_and_create_connect` is on, moves that monitor into a closure `asio::post()`-ed as a plain scheduler task rather than tied to the connection's own socket (`external/restinio/dev/restinio/impl/acceptor.hpp`, `do_accept_current_connection`). If no I/O thread has dequeued it yet, it sits alive in `asio::io_context`'s `scheduler::op_queue_`.
- `RestinioHttpServer::stopAccepting()` (`server->stop(); server->wait();`) only closes the acceptor and joins the I/O threads — it never drains queued/pending operations (`http_server_t::close_sync()` only calls `m_acceptor->close()`; `ioctx_on_thread_pool_t::stop()/wait()` only calls `io_context::stop()` and joins threads).
- `RestinioHttpServer::stop()` then destroys the `ServerHandle`, triggering `~io_context()`. Service shutdown order is fixed (newest-registered service first): the lazily-created `epoll_reactor` shuts down before the always-present `scheduler`. `epoll_reactor::shutdown()` immediately destroys the acceptor's pending "next accept" operation — the last owner of `shared_ptr<acceptor_t>` — freeing the acceptor and its embedded limiter. Only afterward does `scheduler::shutdown()` drain its own queue and reach the still-unexecuted connection-setup closure, whose destructor reads the now-freed limiter.

### Additional detail: exactly how the shutdown path gets into this state

Traced the full call chain from `RestinioHttpServer::stop()` down through RESTinio and asio internals to understand precisely why this isn't just "sometimes things get destroyed in a bad order," but a specific, describable sequence:

1. **`stopAccepting()` only closes the acceptor, nothing else.** `server->stop()` → RESTinio's `on_pool_runner_t::stop()` → `http_server_t::close_async()` → `close_sync()` (`http_server.hpp:372-382`) calls only `m_acceptor->close()`. It never touches a live connection. Once that completes, `on_pool_runner_t`'s `on_ok` callback calls `m_pool.stop()` → `ioctx_on_thread_pool_t::stop()` → `io_context::stop()` (`asio/detail/impl/ioctx_on_thread_pool.hpp:146`).
2. **`io_context::stop()` gives no delivery guarantee.** It makes every thread's `run()` return "as soon as possible" — there's no guarantee any specific pending or just-became-ready operation gets executed first. `server->wait()` only joins those threads; it does not flush the `io_context`.
3. **`~io_context()` tears down whatever is left, and which service runs first is fixed but doesn't line up with the acceptor/connection dependency.** `service_registry::shutdown_services()` (`service_registry.ipp:38-44`) walks services starting from the most-recently-registered one. `scheduler` is created with the `io_context` itself (very early); `epoll_reactor` is created lazily, the first time something needs it (later). So **`epoll_reactor::shutdown()` always runs before `scheduler::shutdown()`.**
4. **`epoll_reactor::shutdown()` destroys the acceptor's pending "next accept" immediately.** It harvests every socket still registered with epoll (`epoll_reactor.ipp:82-99`) — including the acceptor's own long-lived listening socket, whose pending accept operation holds the **last** `shared_ptr<acceptor_t>` — and destroys them all synchronously via `scheduler::abandon_operations()`, whose `op_queue::~op_queue()` really does call `->destroy()` on everything right there (`op_queue.hpp:71-77`). This is where the acceptor (and the `connection_count_limiter` embedded inside it) can actually get freed.
5. **Only afterward does `scheduler::shutdown()` drain whatever was purely *posted*, not tied to any socket** (`scheduler.ipp:178-195`).
6. **Why a connection's cleanup can end up in that second, later-draining bucket at all:** RESTinio's `.separate_accept_and_create_connect(true)` (set at `RestinioHttpServer.cpp:1339`) means `do_accept_current_connection()` (`impl/acceptor.hpp`) constructs the connection's `connection_lifetime_monitor_t` (which holds a **raw pointer** into the acceptor's embedded limiter — by RESTinio's own contract, the acceptor must outlive it) immediately on accept, then moves it into a closure handed to `asio::post()` as a plain scheduler task, not tied to the connection's own socket (that only happens once the closure actually runs `conn->init()`). If no I/O thread dequeues that closure before `stop()` lands, it sits — still alive — purely in `scheduler`'s queue, i.e. exactly the bucket that drains *after* the acceptor has already been freed in step 4.

So the crash isn't really "asio picked an unlucky order once" — it's that step 6 puts one specific kind of object (an accepted-but-not-yet-fully-set-up connection) into the one bucket (step 5) that is structurally guaranteed to be processed *after* the bucket that can free the acceptor it depends on (step 4).

**Comment by matigarciadev** [comment]
## Update 2026-08-31

- Reproduced both test failures by compiling `remoted_module_utest` with ASan and running each affected test repeatedly (`--gtest_repeat`) until they failed.
- Fixed the `FakeUdsServer` issue (Finding 1, stack-use-after-return): tracked and joined its per-connection threads instead of detaching them.
- Analyzed the second case (Finding 2) and fixed the *test*: `AdminBindFailureOnlyWarnsAndTheModuleKeepsServing`'s probe used to open a raw TCP connection and close it immediately, without completing the TLS handshake. I changed it to perform a full TLS handshake and request/response round trip instead, which made the test pass reliably (verified with a high repeat count under ASan).
- The production code behind Finding 2 is still open. Outside of the test suite entirely, using a separate standalone program (not gtest, no test-only headers) that calls the real production entry points (`remoted_module_start`/`remoted_module_stop`) directly, I was able to reproduce the underlying bug in production code as well — confirming it isn't a test-only artifact.
- Analyzed two possible directions for the production fix: (1) turning `separate_accept_and_create_connect` off, and (2) moving to a self-managed `io_context` so shutdown can safely drain pending work before it's destroyed. Both were discussed (tradeoffs, risk, verification approach) but neither has been implemented yet — the production fix remains open.

**Comment by matigarciadev** [comment]
## Update 2026-09-01

- Tried the other production-fix direction first: a self-managed `io_context` with a bounded shutdown grace period, keeping `separate_accept_and_create_connect(true)` so the accept-burst headroom it buys wouldn't be lost. Implemented it, then found it doesn't actually hold up: under `--gtest_repeat` on `ShutdownRace.*`, it hit the *same class* of heap-use-after-free through a different object (a streamed response's `write_group_t`/`connection_lifetime_monitor_t`), because `write_http_response_timelimit`'s per-write timeout resets on every chunk and so never bounds a streamed transfer's total duration — the grace period can elapse while a legitimate download is still in flight, and forcing a stop at that point is exactly as unsafe as the original bug. Reverted this approach.
- Implemented the other direction instead: `separate_accept_and_create_connect(false)`. Verified against the exact historical reproducer at 1000+ cumulative repeats under ASan with zero failures.
- Checked whether this is a known/reported RESTinio issue upstream (searched Stiffstream/restinio's own issue tracker and the web for the relevant symbols/settings) — found nothing matching. Filed it: [Stiffstream/restinio#246](https://github.com/Stiffstream/restinio/issues/246). No response yet.
- While restoring the accept-burst headroom lost by turning that flag off (raising `concurrentAccepts` from a fixed 2 to `cpp_get_nproc()`), found a **separate, pre-existing** issue: at higher accept concurrency, a rare UBSan-flagged null-pointer race surfaces inside asio's `epoll_reactor::start_op()`. Confirmed via isolation testing that this is independent of the `separate_accept_and_create_connect` fix (the fix stays clean at every concurrency level tested) and that it isn't a clean threshold — it's probabilistic and just gets more likely as concurrency rises (0 hits in 1000+ repeats at concurrency 2, 1 hit in ~340 cumulative repeats at concurrency 22, clean runs at 8 and 16). Root cause not fully pinned (would need asio-internal instrumentation to nail down with certainty). Decision: accept this as a known residual risk for now rather than block on it, with a floor of 2 added to the nproc resolution so a single-vCPU host/cgroup can't regress below the old fixed default either way.
- Auxiliary, found while preparing to add `remoted_module` to the repo's clang-format CI workflow: 4 pre-existing formatting violations in that directory (including the exact lambda in `RestinioHttpServer.cpp` that kept reappearing across edits this whole investigation — turns out it really was a checked-in deviation from `.clang-format`, not a local tooling quirk). Fixed those, then added `remoted_module` and `shared_modules/metrics` to the workflow, following its existing per-module pattern.
